File: src/scala/org/codehaus/testdox/intellij/inspections/EmptyTestClassInspection.scala
package org.codehaus.testdox.intellij.inspections

import com.intellij.codeInspection.ProblemHighlightType.GENERIC_ERROR_OR_WARNING
import com.intellij.codeInspection.{InspectionManager, ProblemDescriptor}
import com.intellij.psi.PsiClass

import org.jetbrains.annotations.NotNull
import org.codehaus.testdox.intellij.TestDoxProjectComponent
import org.codehaus.testdox.intellij.inspections.Inspection.NO_PROBLEMS

class EmptyTestClassInspection extends Inspection {

  @NotNull
  val getDisplayName = "Test class with no tests"

  override def checkClass(@NotNull psiClass: PsiClass, @NotNull manager: InspectionManager, isOnTheFly: Boolean): Array[ProblemDescriptor] = {
    if (isInnerClass(psiClass) || isSuite(psiClass)) {
      return NO_PROBLEMS
    }

    val testDoxController = TestDoxProjectComponent.getInstance(manager.getProject()).getController()
    val factory = testDoxController.getTestDoxFileFactory()
    val file = factory.getTestDoxFile(psiClass.getContainingFile().getVirtualFile())

    if (!file.isTestedClass && file.canNavigateToTestedClass && file.testMethods.length == 0)
      Array(manager.createProblemDescriptor(psiClass.getNameIdentifier(), getDisplayName, new AddTestMethodQuickFix(), GENERIC_ERROR_OR_WARNING, false))
    else
      NO_PROBLEMS
  }

  private def isSuite(psiClass: PsiClass) = psiClass.findMethodsByName("suite", true).length > 0

  private def isInnerClass(psiClass: PsiClass) = {
    psiClass.getName() != null && psiClass.getContainingFile().getName().indexOf(psiClass.getName()) < 0
  }
}

File: src/java/org/codehaus/testdox/intellij/inspections/EmptyTestClassInspection.java
package org.codehaus.testdox.intellij.inspections;

import com.intellij.codeInspection.InspectionManager;
import com.intellij.codeInspection.ProblemDescriptor;
import com.intellij.codeInspection.ProblemHighlightType;
import com.intellij.psi.PsiClass;
import static jedi.functional.Coercions.array;
import org.codehaus.testdox.intellij.TestDoxController;
import org.codehaus.testdox.intellij.TestDoxFile;
import org.codehaus.testdox.intellij.TestDoxFileFactory;
import org.codehaus.testdox.intellij.TestDoxProjectComponent;
import org.jetbrains.annotations.NotNull;

public class EmptyTestClassInspection extends AbstractTestDoxInspection {

    @NotNull
    public String getDisplayName() {
        return "Test class with no tests";
    }

    public ProblemDescriptor[] checkClass(@NotNull PsiClass psiClass, @NotNull InspectionManager manager, boolean isOnTheFly) {
        if (isInnerClass(psiClass) || isSuite(psiClass)) {
            return NO_PROBLEMS;
        }

        TestDoxController testDoxController = TestDoxProjectComponent.getInstance(manager.getProject()).getController();
        TestDoxFileFactory factory = testDoxController.getTestDoxFileFactory();
        TestDoxFile file = factory.getTestDoxFile(psiClass.getContainingFile().getVirtualFile());

        if ((!file.isTestedClass()) && (file.canNavigateToTestedClass()) && (file.getTestMethods().length == 0)) {
            return array(
                manager.createProblemDescriptor(
                    psiClass.getNameIdentifier(), getDisplayName(), new AddTestMethodQuickFix(), ProblemHighlightType.GENERIC_ERROR_OR_WARNING)
            );
        }

        return NO_PROBLEMS;
    }

    private boolean isSuite(PsiClass psiClass) {
        return (psiClass.findMethodsByName("suite", true).length > 0);
    }

    private boolean isInnerClass(PsiClass psiClass) {
        return (psiClass.getName() != null) && (psiClass.getContainingFile().getName().indexOf(psiClass.getName()) < 0);
    }
}
